### PR TITLE
[Merged by Bors] - feat: port AlgebraicGeometry.OpenImmersion.Basic

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

- [x] depends on: #5047 (relaxing universe level)

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#5047
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#5047~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,1204 @@
...
+
+/-- The composition of two open immersions is an open immersion. -/
+instance comp {Z : PresheafedSpace C} (f : X ⟶ Y) [hf : is_open_immersion f] (g : Y ⟶ Z)
+    [hg : is_open_immersion g] : is_open_immersion (f ≫ g) where
+  base_open := hg.base_open.comp hf.base_open
+  c_iso U := by
+    generalize_proofs h
+    dsimp only [AlgebraicGeometry.PresheafedSpace.comp_c_app, unop_op, Functor.op, comp_base,
+      TopCat.Presheaf.pushforwardObj_obj, Opens.map_comp_obj]
+    -- Porting note : was `apply (config := { instances := False }) ...`
```
Comment (semorrison) [inline Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean:164]:
```suggestion
    -- Porting note : was `apply (config := { instances := False }) ...`
    -- See https://github.com/leanprover/lean4/issues/2273
```

Comment (semorrison) [reply Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean:164]: (And below.)

File: Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,1204 @@
...
+        erw [s.snd.c.naturality_assoc]
+        rw [Category.assoc]
+        erw [← s.pt.presheaf.map_comp, ← s.pt.presheaf.map_comp]
+        congr 1 }
+#align algebraic_geometry.PresheafedSpace.is_open_immersion.pullback_cone_of_left_lift AlgebraicGeometry.PresheafedSpace.IsOpenImmersion.pullbackConeOfLeftLift
+
+-- this lemma is not a `simp` lemma, because it is an implementation detail
+theorem pullbackConeOfLeftLift_fst :
+    pullbackConeOfLeftLift f g s ≫ (pullbackConeOfLeft f g).fst = s.fst := by
+  -- Porting note : `ext` did not pick up `NatTrans.ext`
```
Comment (semorrison) [inline Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean:448]: Could you add the missing `ext` lemma?

File: Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,1204 @@
...
+#align algebraic_geometry.SheafedSpace.is_open_immersion.SheafedSpace_pullback_to_base_is_open_immersion AlgebraicGeometry.SheafedSpace.IsOpenImmersion.sheafedSpace_pullback_to_base_isOpenImmersion
+
+end Pullback
+
+section OfStalkIso
+
+variable [HasLimits C] [HasColimits C] [ConcreteCategory C]
+
+-- Porting note : `forget ↦ CategoryTheory.forget` because `forget` on its own is
+-- `AlgebraicGeometry.SheafedSpace.forget`
```
Comment (semorrison) [inline Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean:828]:
```suggestion
```

File: Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean (hunk 2 above)
Comment (jjaassoonn) [reply Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean:451]: I tired `attribute [ext] NatTrans.ext`, still does not work. After quick searching porting note, `ext` regression is not an isolated case. @semorrison

**Comment by jcommelin** [comment]
This is currently hard to review, because I can't see the diff between mathport output and the HEAD of the PR.

**Comment by Ruben-VandeVelde** [comment]
> This is currently hard to review, because I can't see the diff between mathport output and the HEAD of the PR.

Should be better now

**Comment by jjaassoonn** [comment]
> > This is currently hard to review, because I can't see the diff between mathport output and the HEAD of the PR.
> 
> Should be better now

Something broke now

**Comment by jjaassoonn** [comment]
@Ruben-VandeVelde, I fixed the broken universe level, can you do the same git magic again (if necessary) to make this easier to review please

**Comment by Ruben-VandeVelde** [comment]
Done, I wonder what happened there...

File: Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean:451]: The issue is that current `ext` doesn't see through type synonyms. I've proposed a fix but it's unclear if we'll actually implement it, so in the meantime the solution is to add an extra `ext` lemma, essentially for the morphisms of every new category that gets defined. Does that make sense?

Comment (semorrison) [reply Mathlib/AlgebraicGeometry/OpenImmersion/Basic.lean:451]: See https://github.com/leanprover-community/mathlib4/pull/5191

**Pull request closed without merging**